pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/MmapMemory.java

```diff
@@ -0,0 +1,348 @@
+  private static final MapFun MAP_FUN;
+
+  /**
+   * The address actually mapped. It has to be page aligned.
+   *
+   * {@code _address = _offset - offset % pageSize}
+   */
+  private final long _address;
+  /**
+   * The offset requested to map.
+   *
+   * {@code _address = _offset - offset % pageSize}
+   */
+  private final long _offset;
+  /**
+   * How many bytes have been requested to be mapped.
+   * The actual mapped size may be larger (up to the next page), but the actual mapped size
+   * is stored by {@link #_section}.
+   */
+  private final long _size;
+  private final MapSection _section;
+
+  static {
+    try {
+      Jvm.init();
+      MAP_FUN = MapFun.find();
+    } catch (ClassNotFoundException | NoSuchMethodException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public MmapMemory(File file, boolean readOnly, long offset, long size) {
+    _size = size;
+    _offset = offset;
+
+    try {
+      _section = MAP_FUN.map(file, readOnly, offset, size);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    _address = _section.getAddress();
+  }
+
+  @Override
+  public long getAddress() {
+    return _address;
+  }
+
+  @Override
+  public long getSize() {
+    return _size;
+  }
+
+  @Override
+  public void flush() {
+    MSyncFlag mode = MSyncFlag.MS_SYNC;
+    PosixAPI.posix().msync(_offset, _size, mode);
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    try {
+      _section._unmapFun.unmap();
+    } catch (InvocationTargetException | IllegalAccessException e) {
+      throw new RuntimeException("Error while calling unmap", e);
+    }
+  }
+
+  private static class MapSection {
+    public static final MapSection EMPTY = new MapSection(0, () -> {
+    });
+    private final long _address;
+    private final UnmapFun _unmapFun;
+
+    public MapSection(long address, UnmapFun unmapFun) {
+      _address = address;
+      _unmapFun = unmapFun;
+    }
+
+    public long getAddress() {
+      return _address;
+    }
+
+    public UnmapFun getUnmapFun() {
+      return _unmapFun;
+    }
+  }
+
+  interface MapFun {
+
+    /**
+     * @param file The file to be mapped. If its length is lower than offset + size and the mode is not read only,
+     *            the file will be resized to that size.
+     * @param offset The offset in the file. Any positive value is valid, even if it is larger than the file size.
+     * @param size How many bytes to map.
+     * @throws IOException in several situations. For example, if the offset + size is larger than file length and the
+     * mode is read only or if the process doesn't have permission to read or modify the file.
+     */
+    MapSection map(File file, boolean readOnly, long offset, long size) throws IOException;
+
+    static MapFun find()
+        throws ClassNotFoundException, NoSuchMethodException {
+      List<Finder<? extends MapFun>> candidates = Lists.newArrayList(
+          new Map0Fun.ChronicleCore(),
```

Comment:
With this ChronicleCore library we could remove all other options. They are already doing more or less the same thing we do here (taking care of more details and with more expertise than I have). We should decide whether we want to use their library (and remove `Map0Fun.Java11`, `Map0Fun.Java17` and `Java20`) or reinvent the wheel with the other `MapFun`.

Reply:
another question - previous experiments / attempts by you had concluded that Chronicle Bytes has several blocking bugs and was not suitable for use by us. What changed since then that we are back to using it here and also preferring it to be the default option?
I thought that option 4 was the obvious choice, but seems like I'm missing something

Reply:
> shouldn't the options after ChronicleCore be in reverse order i.e. Java20 first, followed by 17 then 11?

Option 2 is prioritize Chronicle Core. Therefore it should be the first one. If that one fails (the internal methods cannot be found) then it will try to use my code. Option 3 would be to place Chronicle Core at the end, so it will be used as fallback

>  If we're running on Java 20, won't the older java methods and classes still be found?

No. The reason why we need to add different ways to look for the method is that these methods are removed/renamed/changed on each JVM. So the method used in Java 8 will not be found in Java 20 (and the other way around).

The JVM can break compatibility with these methods because they are private and therefore by changing them they do not break backward compatibility.

Reply:
I meant, in option 2, the order should be Chronicle -> Java 20 -> Java 17 -> Java 11 instead of Chronicle -> java 11 -> Java 17 -> Java 20 ? Or is the function name + signature guaranteed unique across versions?

Reply:
AFAIK they are unique across different JVM versions.